### PR TITLE
feat(x/ecocredit): add curation tag API

### PR DESCRIPTION
**Pull request by aaronc**
## Description

Ref: #374, replaces #395 

Adds curation tags to the x/ecocredit data model which can be scoped to a class, project (depends on #608) or batch.

The motivation for adding tags directly to x/ecocredit is to facilitate an order book based marketplace which can leverage tags. Having all of this metadata local to the module will facilitate making the orderbook scalable and performant.

The focus of this approach compared to #395 as much simplicity as possible. There is no explicit registration of tags, rather they are very simply `{curator}/{tag_name}` (ex. `regen1358hg239812h4usdg9/top-quality-credit`) and cannot be transferred between different curators. I am not sure of the pros/cons of this model but it seems like maybe it is an okay simplication because we mostly want curation from long-lived entities like DAOs. But if transfer between curators is desired we can add that back in. Also, there is no explicit on-chain tag metadata here with the thinking that this could be handled off-chain.

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [ ] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] added `!` to the type prefix if API or client breaking change
- [ ] targeted the correct branch (see [PR Targeting](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] provided a link to the relevant issue or specification
- [ ] followed the guidelines for [building modules](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules)
- [ ] included the necessary unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] added a changelog entry to `CHANGELOG.md`
- [ ] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed 
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: proto/regen/ecocredit/v1alpha1/types.proto [hunk 1]
```diff
@@ -117,4 +117,18 @@ message CreditTypeSeq {
...
+// TagTarget represents the target of a tag.
+message TagTarget {
+  oneof sum {
+    // class_id indicates that this tag targets all the credits within a given class.
+    string class_id = 1;
+
+    // project_id indicates that this tag targets all the credits within a given project.
+    string project_id = 2;
+
+    // project_id indicates that this tag targets the credits within a batch.
```
Comment (technicallyty) [inline proto/regen/ecocredit/v1alpha1/types.proto:131]:
```suggestion
    // batch_id indicates that this tag targets the credits within a batch.
```

**Review by technicallyty — COMMENTED** [review]
should we also include a query for getting all the tags from a specific curator? could be a nice inverse from the current query -> as a way to see a specific curator's tags.

**Comment by aaronc** [comment]
> should we also include a query for getting all the tags from a specific curator? could be a nice inverse from the current query -> as a way to see a specific curator's tags.

Sure 👍

File: proto/regen/ecocredit/v1alpha1/query.proto [hunk 2]
```diff
@@ -172,8 +175,28 @@ message QuerySupplyResponse {
...
 }
+
+// QueryTagsRequest is the Query/Tags request type
+message QueryTagsRequest {
+
+  // target is the tag target to query.
+  TagTarget target = 1;
+
+  // pagination defines an optional pagination for the request.
+  cosmos.base.query.v1beta1.PageRequest pagination = 1;
```
Comment (ryanchristo) [inline proto/regen/ecocredit/v1alpha1/query.proto:191]:
```suggestion
  cosmos.base.query.v1beta1.PageRequest pagination = 2;
```

File: proto/regen/ecocredit/v1alpha1/tx.proto [hunk 3]
```diff
@@ -427,3 +431,29 @@ message MsgAllowAskDenom {
...
+  // Tagging is a tag operations.
+  message Tagging {
+    // tag_name is the name of the tag to apply to this target. It is scoped to the curator
+    // which is applying the tag, so the resultant tag is effectively {curator}/{tag_name}.
+    string tag_name = 1;
+
+    // targets are the tag operation targets.
+    repeated TagTarget targets = 2;
+
+    // untag indicates the a tag should be removed if set to true.
```
Comment (blushi) [inline proto/regen/ecocredit/v1alpha1/tx.proto:452]:
```suggestion
    // untag indicates that a tag should be removed if set to true.
```

File: proto/regen/ecocredit/v1alpha1/types.proto [hunk 4]
```diff
@@ -117,4 +117,18 @@ message CreditTypeSeq {
 
   // The sequence number of classes of the credit type
   uint64 seq_number = 2;
-}
\ No newline at end of file
+}
+
+// TagTarget represents the target of a tag.
+message TagTarget {
+  oneof sum {
```
Comment (blushi) [inline proto/regen/ecocredit/v1alpha1/types.proto:124]: would that make sense to also be able to target credit types?

File: proto/regen/ecocredit/v1alpha1/types.proto [hunk 5]
```diff
@@ -117,4 +117,18 @@ message CreditTypeSeq {
...
+    // class_id indicates that this tag targets all the credits within a given class.
+    string class_id = 1;
+
+    // project_id indicates that this tag targets all the credits within a given project.
+    string project_id = 2;
+
+    // batch_id indicates that this tag targets the credits within a batch.
+    string batch_id = 3;
```
Comment (blushi) [inline proto/regen/ecocredit/v1alpha1/types.proto:132]:
```suggestion
    // batch_denom indicates that this tag targets the credits within a batch.
    string batch_denom = 3;
```

File: proto/regen/ecocredit/v1alpha1/types.proto (hunk 4 above)
Comment (aaronc) [reply proto/regen/ecocredit/v1alpha1/types.proto:124]: Credit types should be so few that I'm not sure there's a value... @clevinson ?

**Comment by aaronc** [comment]
Closing in favor of #705

**Pull request closed without merging**